File: 2018/day14/day14.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *board = NULL;
int boardLen = 0;
int bp = 0;  /* next location to add a number */
char part1[11] = { -1 };
int part2 = -1;
char *input = "540561";
char inputArray[1024];
int inputVal;
int inputLen;

void addRecipe( char val )
{
  if(bp >= boardLen) {
    /* Allocate more space for board. */
    boardLen = 1024 + boardLen * 2;
    board = realloc(board, boardLen);
  }
  board[bp++] = val;

  if(bp == inputVal + 10) {
    memcpy(part1, board + bp - 10, 10);
    part1[10] = '\0';
  }

  if(part2 < 0 && bp >= inputLen) {
    if(!memcmp(board + bp - inputLen, inputArray, inputLen)) {
      part2 = bp - inputLen;
    }
  }
}

int main( int argc, char **argv )
{
  int elf1 = 0, elf2 = 1;

  if(argc > 1) input = argv[1];
  inputVal = atoi(input);
  inputLen = strlen(input);
  for(int i = 0; i < inputLen; i++) {
    inputArray[i] = input[i] - '0';
  }

  addRecipe(3);
  addRecipe(7);

  while(part1[0] < 0 || part2 < 0) {
    int newRecipe = board[elf1] + board[elf2];
    if(newRecipe > 9) {
      addRecipe(1);
      newRecipe -= 10;
    }
    addRecipe(newRecipe);

    elf1 = (elf1 + 1 + board[elf1]) % bp;
    elf2 = (elf2 + 1 + board[elf2]) % bp;
  }

  /* convert part 1 result to printable digits */
  for(int i = 0; i < 10; i++) part1[i] += '0';
  printf("Part 1 result: %s\nPart 2 result: %d\n", part1, part2);
  return 0;
}

File: 2018/day14/day14.pl
use strict vars;

our $input = shift // 540561;
our $firstDigit = substr($input, 0, 1);

our @board = (3, 7);
our $elf1 = 0;  # elf 1 pointer
our $elf2 = 1;  # elf 2 pointer
our $part1 = -1;
our $part2 = -1;

while($part1 < 0 || $part2 < 0) {
  my $newRecipe = $board[$elf1] + $board[$elf2];
  # Add the 10s digit if required (has to be 1 or 0) and then add the ones
  # digit.
  if($newRecipe > 9) {
    addRecipe(1);
    $newRecipe -= 10;
  }
  addRecipe($newRecipe);
  
  # move elf pointers
  $elf1 = ($elf1 + 1 + $board[$elf1]) % scalar(@board);
  $elf2 = ($elf2 + 1 + $board[$elf2]) % scalar(@board);
}

print "Part 1 result: $part1\n";
print "Part 2 result: $part2\n";

sub addRecipe {
  push @board, $_[0];

  # part 1 -- scores of the ten recipes right after the input number
  if(scalar @board == $input + 10) {
    $part1 = join '', @board[-10 .. -1];
  }

  # part 2 -- location of the input sequence on the board
  if($part2 < 0) {
    # figure out if the sequence ends with the digits from input
    my $len = length $input;
    # The first digit check is just to quickly eliminate possibilities
    # before doing the relatively slow process of joining the last few
    # digits of @board.
    if($board[-$len] == $firstDigit &&
       $input eq join('', @board[-$len .. -1])) {
      # Result is the current length of sequence, minus the length of the
      # input.
      $part2 = scalar @board - $len;
    }
  }
}
